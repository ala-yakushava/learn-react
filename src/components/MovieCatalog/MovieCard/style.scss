@import "../../../styles/theme";

.MovieCard {
  position: relative;
  display: flex;
  flex-wrap: wrap;
  align-items: flex-start;

  &_picture {
    display: block;
    width: 100%;
    height: 456px;
    margin-bottom: 20px;
    background-color: $grey;
  }

  &_title {
    margin: 0;
    margin-bottom: 3px;
    font-size: 22px;
    line-height: 1.2;
    color: $white;
  }

  &_year {
    margin-left: auto;
    padding: 2px 15px;
    font-size: 16px;
    line-height: 1.2;
    color: $white;
    border: 1px solid $white;
    border-radius: 5px;
  }

  &_genre-list {
    width: 100%;
    margin: 0;
    font-size: 16px;
    line-height: 1.2;
    color: $white;
  }

  &_menu {
    position: absolute;
    top: 10px;
    right: 10px;
  }

  &_button {
    @include reset-button;
    width: 40px;
    height: 40px;
    margin-left: auto;
    font-size: 30px;
    color: $white;
    border: none;
    background-color: $darkest-grey;
    border-radius: 50%;
    transform: rotate(90deg);

    span {
      position: relative;
      top: -8px;
    }
  }
}
