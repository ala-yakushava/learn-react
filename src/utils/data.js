import uniqueId from 'lodash/uniqueId';

export const sortItems = [
  {
    id: 'sort-rating',
    value: 'vote_average',
    label: 'Rating',
  },
  {
    id: 'sort-release_date',
    value: 'release_date',
    label: 'Release date',
  },
];

export const filterItems = [
  {
    id: 'filter-all',
    value: 'All',
    label: 'All',
  },
  {
    id: 'filter-documentary',
    value: 'Documentary',
    label: 'Documentary',
  },
  {
    id: 'filter-comedy',
    value: 'Comedy',
    label: 'Comedy',
  },
  {
    id: 'filter-horror',
    value: 'Horror',
    label: 'Horror',
  },
  {
    id: 'filter-crime',
    value: 'Crime',
    label: 'Crime',
  },
];

export const formField = [
  {
    id: uniqueId(),
    name: 'title',
    type: 'text',
    label: 'Title',
    placeholder: 'Title here',
  },
  {
    id: uniqueId(),
    name: 'release_date',
    type: 'date',
    label: 'Release date',
    placeholder: 'Select date',
  },
  {
    id: uniqueId(),
    name: 'poster_path',
    type: 'url',
    label: 'Movie URL',
    placeholder: 'Movie URL here',
  },
  {
    id: uniqueId(),
    name: 'genres',
    type: 'text',
    label: 'Genre',
    placeholder: 'Select genre',
  },
  {
    id: uniqueId(),
    name: 'overview',
    type: 'text',
    label: 'Overview',
    placeholder: 'Overview text goes here',
  },
  {
    id: uniqueId(),
    name: 'runtime',
    type: 'text',
    label: 'Runtime',
    placeholder: 'Runtime text goes here',
  },
];
