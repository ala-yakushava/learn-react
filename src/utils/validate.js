import * as yup from 'yup';

const MovieFormSchema = yup.object().shape({
  title: yup.string().transform((value) => value.trim()).min(2).required('Required'),
  release_date: yup.date().max(new Date(), 'Future date'),
  poster_path: yup.string().url('Invalid URL').required('Required'),
  genres: yup.string().transform((value) => value.trim()).min(3).required('Required'),
  overview: yup.string().transform((value) => value.trim()).min(10).required('Required'),
  runtime: yup.number('Invalid number').required('Required').min(0, 'Too Short!').integer(),
});

export default MovieFormSchema;
