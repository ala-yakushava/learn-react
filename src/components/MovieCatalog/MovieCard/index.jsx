import React, { useState } from 'react';
import PropTypes from 'prop-types';

import './style.scss';
import MovieCardMenu from '../MovieCardMenu';

const MovieCard = ({ movie, onClick, onRemoveMovie }) => {
  const [visible, setVisible] = useState(false);

  const {
    id, title, release_date: date, poster_path: poster, genres,
  } = movie;

  const handleClickOpenMenu = (evt) => {
    evt.stopPropagation();
    setVisible(true);
  };

  const handleClickCloseMenu = (evt) => {
    evt.stopPropagation();
    setVisible(false);
  };

  return (
    // eslint-disable-next-line
    <article className="MovieCard" onClick={onClick}>
      <img
        className="MovieCard_picture"
        src={poster}
        alt={`Poster for the movie ${title}`}
      />
      <h2 className="MovieCard_title">{ title }</h2>
      <span className="MovieCard_year">{ date }</span>
      <p className="MovieCard_genre-list">{ genres.join(', ') }</p>
      <div className="MovieCard_menu">
        <button type="button" className="MovieCard_button" onClick={handleClickOpenMenu}>
          <span>...</span>
        </button>
        {visible && (
          <MovieCardMenu
            movieId={id}
            onClickCloseMenu={handleClickCloseMenu}
            onRemoveMovie={onRemoveMovie}
          />
        )}
      </div>
    </article>
  );
};

MovieCard.propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    genres: PropTypes.arrayOf(PropTypes.string).isRequired,
    release_date: PropTypes.string.isRequired,
    poster_path: PropTypes.string.isRequired,
  }).isRequired,
  onClick: PropTypes.func.isRequired,
  onRemoveMovie: PropTypes.func.isRequired,
};

export default MovieCard;
