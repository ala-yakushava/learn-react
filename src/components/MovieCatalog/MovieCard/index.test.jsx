import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import MovieCard from './index';
import { movie } from '../../../utils/mock-data';

describe('Components - MovieCard', () => {
  const handleRemove = jest.fn();

  test('should match snapshot', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <MovieCard movie={movie} onRemoveMovie={handleRemove} />
        </MemoryRouter>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should open MovieCardMenu', () => {
    const { getByTestId, getByRole } = render(
      <MemoryRouter>
        <MovieCard movie={movie} onRemoveMovie={handleRemove} />
      </MemoryRouter>,
    );

    const leftClick = { button: 0 };
    const button = getByRole('button', { name: /.../i });
    userEvent.click(button, leftClick);

    const menu = getByTestId('card-menu');
    expect(menu).toBeInTheDocument();

    const closeMenu = getByRole('button', { name: /Ñ…/i });
    userEvent.click(closeMenu, leftClick);

    expect(menu).not.toBeInTheDocument();
  });
});
