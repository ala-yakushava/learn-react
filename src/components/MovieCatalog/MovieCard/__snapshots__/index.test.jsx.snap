// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - MovieCard should match snapshot 1`] = `
<article
  className="MovieCard"
  data-testid="movie-card"
>
  <a
    aria-current={null}
    href="/movies/284053"
    onClick={[Function]}
  >
    <img
      alt="Poster for the movie Thor: Ragnarok"
      className="MovieCard_picture"
      src="https://image.tmdb.org/t/p/w500/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg"
    />
  </a>
  <h2
    className="MovieCard_title"
  >
    Thor: Ragnarok
  </h2>
  <span
    className="MovieCard_year"
  >
    2017-10-25
  </span>
  <p
    className="MovieCard_genre-list"
  >
    Action, Adventure, Fantasy
  </p>
  <div
    className="MovieCard_menu"
  >
    <button
      className="MovieCard_button"
      onClick={[Function]}
      type="button"
    >
      <span>
        ...
      </span>
    </button>
  </div>
</article>
`;
